akka-typed/src/main/scala/akka/typed/cluster/ClusterSingleton.scala

```diff
@@ -0,0 +1,185 @@
+/**
+ * Copyright (C) 2009-2017 Lightbend Inc. <http://www.lightbend.com>
+ */
+package akka.typed.cluster
+
+import akka.actor.{ ExtendedActorSystem, InvalidActorNameException, NoSerializationVerificationNeeded }
+import akka.annotation.{ DoNotInherit, InternalApi }
+import akka.cluster.ClusterSettings.DataCenter
+import akka.cluster.singleton.{ ClusterSingletonManagerSettings, ClusterSingletonProxy, ClusterSingletonProxySettings, ClusterSingletonManager ⇒ OldSingletonManager }
+import akka.typed.internal.adapter.{ ActorRefAdapter, ActorSystemAdapter }
+import akka.typed.scaladsl.adapter._
+import akka.typed.{ ActorRef, ActorSystem, Behavior, Extension, ExtensionId, Props }
+import com.typesafe.config.Config
+
+import scala.concurrent.duration.FiniteDuration
+
+object ClusterSingletonSettings {
+  def apply(
+    system: ActorSystem[_]
+  ): ClusterSingletonSettings = fromConfig(system.settings.config.getConfig("akka.cluster"))
+
+  def fromConfig(
+    config: Config
+  ): ClusterSingletonSettings = {
+    // TODO introduce a config namespace for typed singleton and read that?
+    // currently singleton name is required and then discarded, for example
+    val mgrSettings = ClusterSingletonManagerSettings(config.getConfig("singleton"))
+    val proxySettings = ClusterSingletonProxySettings(config.getConfig("singleton-proxy"))
+    new ClusterSingletonSettings(
+      mgrSettings.role,
+      proxySettings.dataCenter,
+      proxySettings.singletonIdentificationInterval,
+      mgrSettings.removalMargin,
+      mgrSettings.handOverRetryInterval,
+      proxySettings.bufferSize
+    )
+  }
+}
+
+final class ClusterSingletonSettings(
+  val role:                            Option[String],
+  val dataCenter:                      Option[DataCenter],
+  val singletonIdentificationInterval: FiniteDuration,
+  val removalMargin:                   FiniteDuration,
+  val handOverRetryInterval:           FiniteDuration,
+  val bufferSize:                      Int) extends NoSerializationVerificationNeeded {
+
+  /**
+   * Replace the required roles with a new set of required roles
+   */
+  def withRole(role: String): ClusterSingletonSettings =
+    new ClusterSingletonSettings(Some(role), dataCenter, singletonIdentificationInterval, removalMargin, handOverRetryInterval, bufferSize)
```

Comment:
isn't it better to have one private `copy` method rather than repeating all args in all these places?

Reply:
Good point.

Reply:
You can use github.com/ktoso/kaze-class to generate the style we have in most places - private `copy` and `withXXX` methods

Reply:
I'd want something like that, but I don't want to introduce something like that unless we make a team decision to go with it.

Reply:
I'm positive to using kaze-class as initial bootstrapping of such settings classes, assuming that it creates the things we would otherwise anyway write manually.

Reply:
It's a tool that just prints stuff that you then copy-paste into your code ;) Entire akka-http settings were initially made like this, then adding settings is a manual thing later on, but simple because one-by-one

Reply:
Yeah, mostly for bootstrap of boilerplate heavy classes like that.
It generates this: https://github.com/akka/akka/pull/23676/files#diff-65d0185a010d982ad5b37c2cc9298c38R69 

hidden constructor (people should not use it, we can internally), and `with...` methods, the copy method is available but hidden, again, has to be; it does also toString, but not hashcode/equals, though here i dont think we care about those

Reply:
👍